apiserver/facades/client/application/application.go

```diff
@@ -277,6 +277,16 @@ func (api *APIBase) Deploy(args params.ApplicationsDeploy) (params.ErrorResults,
 	}
 
 	for i, arg := range args.Applications {
+		// If either the charm origin ID or Hash is set before a charm is
+		// downloaded, charm download will fail for charms with a forced series.
+		// The logic (refreshConfig) in sending the correct request to charmhub
+		// will break.
+		if arg.CharmOrigin != nil &&
+			((arg.CharmOrigin.ID != "" && arg.CharmOrigin.Hash == "") ||
+				(arg.CharmOrigin.ID == "" && arg.CharmOrigin.Hash != "")) {
```

Comment:
Nit: you can remove the redundant parentheses since `&&` is higher precedence than `||` (also in a couple of places below):

```go
	if arg.CharmOrigin != nil &&
		(arg.CharmOrigin.ID != "" && arg.CharmOrigin.Hash == "" ||
			arg.CharmOrigin.ID == "" && arg.CharmOrigin.Hash != "") {
		...
	}
```

Reply:
I find it easier to read with the added parentheses, neither the formatter nor lint are complaining.